### PR TITLE
LibDSP+Piano: Implement a band adjustment filter on the road to equalizers

### PR DESCRIPTION
**Pull request by kleinesfilmroellchen**
Band adjustment filters are a fundamental building block of equalizers and similar effects. This PR not only implements the fundamental abstractions in LibDSP, but also makes the filter usable through a basic effect in Piano.

This also includes a necessary bugfix for temporary sample aliasing, which this PR exposed.

### CODE REVIEW

File: Userland/Libraries/LibDSP/BandAdjustmentFilter.cpp [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+    // G0 and G1
+    auto constexpr offset_gain = 1.f;
+
+    // Ω0, eq. 19
+    auto const analog_center_frequency = AK::tan(digital_center_frequency_radians / 2);
+    // W², eq. 18
+    auto const w_squared = AK::sqrt((m_gain * m_gain - offset_gain * offset_gain) / (m_gain * m_gain - offset_gain * offset_gain)) * analog_center_frequency * analog_center_frequency;
+    // ΔΩ, eq. 19
+    auto const analog_bandwidth = (1 + AK::sqrt((bandwidth_gain * bandwidth_gain - offset_gain * offset_gain) / (bandwidth_gain * bandwidth_gain - offset_gain * offset_gain)) * w_squared) * AK::tan(digital_bandwidth_radians / 2.0f);
+    // eq. 17
```
Comment (konradekk) [inline Userland/Libraries/LibDSP/BandAdjustmentFilter.cpp:62]: it seems you refer to some paper or other reference but itʼs nowhere to be mentioned here… 🤔 (at least at first glance!)

care to add it to this file? 🙇🏻‍♂️

Comment (konradekk) [reply Userland/Libraries/LibDSP/BandAdjustmentFilter.cpp:62]: oh! neat – thx! 🙇🏻‍♂️

File: Userland/Libraries/LibAudio/Sample.h [hunk 2]
```diff
@@ -173,6 +179,13 @@ struct Sample {
         return *this;
     }
 
+    constexpr Sample& operator-=(Sample const& other)
+    {
+        left -= other.left;
+        right -= other.right;
+        return *this;
+    }
```
Comment (GMTA) [inline Userland/Libraries/LibAudio/Sample.h:199]: Why have two commits adding generic operators to `Sample`? I'd combine this with the first commit (the one adding multiply/divide)

File: Userland/Libraries/LibAudio/Sample.h [hunk 3]
```diff
@@ -45,6 +46,12 @@ struct Sample {
     constexpr Sample(Sample&& other) = default;
     constexpr Sample& operator=(Sample&& other) = default;
 
+    // *Either* channel has a NaN value.
+    bool is_nan() const
```
Comment (GMTA) [inline Userland/Libraries/LibAudio/Sample.h:55]: If you have to add a comment to explain what a function does, maybe the function could explain itself better by its name?

E.g. `either_channel_is_nan()` or `any_channel_is_nan()`

File: Userland/Libraries/LibAudio/Sample.h [hunk 4]
```diff
@@ -46,6 +46,11 @@ struct Sample {
     constexpr Sample(Sample&& other) = default;
     constexpr Sample& operator=(Sample&& other) = default;
 
+    constexpr bool operator==(Sample& other) const
+    {
+        return this->left == other.left && this->right == other.right;
+    }
```
Comment (GMTA) [inline Userland/Libraries/LibAudio/Sample.h:52]: This {c/sh}ould also go into the first commit

File: Userland/Libraries/LibDSP/Effects.cpp [hunk 5]
```diff
@@ -89,4 +89,37 @@ void Mastering::process_to_fixed_array(Signal const& input_signal, FixedArray<Sa
...
+    m_parameters.append(m_q);
+
+    m_frequency.register_change_listener([this](auto const& frequency) {
+        m_filter.set_center_frequency(static_cast<double>(frequency));
+    });
+    m_gain_db.register_change_listener([this](auto const& gain_db) {
+        m_filter.set_gain_db(static_cast<double>(gain_db));
+    });
+    m_q.register_change_listener([this](auto const& q) {
+        m_filter.set_q(static_cast<double>(q));
```
Comment (GMTA) [inline Userland/Libraries/LibDSP/Effects.cpp:110]: Why these casts to `double` when the argument takes in a `float`?

File: Userland/Libraries/LibAudio/Sample.h (hunk 2 above)
Comment (kleinesfilmroellchen) [reply Userland/Libraries/LibAudio/Sample.h:199]: Right, I'd rather do that than to split the commit up after the fact :^)

File: Userland/Libraries/LibDSP/Effects.cpp (hunk 5 above)
Comment (kleinesfilmroellchen) [reply Userland/Libraries/LibDSP/Effects.cpp:110]: Copy-paste, since we need to create a floating-point value from the fixed-point parameters.

**Comment by kleinesfilmroellchen** [comment]
This filter (or the "improved" version that's on the parametric EQ branch) is currently not quite correct; I will need to do more signal processing debugging to make it work correctly with both positive and negative gain.

**Pull request closed without merging**